src/App.js

```diff
@@ -30,36 +30,91 @@ const App = () => {
   const [squares, setSquares] = useState(generateSquares());
 
-  // Wave 2
-  // You will need to create a method to change the square
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+  //update current player
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+
+  // sets the game winner
+  const [gameWinner, setGameWinner] = useState('');
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
   };
 
   const resetGame = () => {
-    // Complete in Wave 4
+    setSquares(generateSquares());
+  };
+
+  const onClickCallback = (id) => {
+    console.log('click', id);
+    const newSquares = [...squares];
+
+    // returns early if a winner has been choosen
+    if (checkForWinner()) {
+      return;
+    }
+
+    for (let i = 0; i < newSquares.length; i++) {
+      for (let j = 0; j < newSquares[i].length; j++) {
+        if (newSquares[i][j].id === id && newSquares[i][j].value === '') {
+          newSquares[i][j].value = currentPlayer;
+          // if square has been clicked this
+          // prevents it from being clicked again
+        } else if (newSquares[i][j].id === id) {
+          return;
+        }
+      }
+    }
+    setSquares(newSquares);
+    setCurrentPlayer(currentPlayer === PLAYER_1 ? PLAYER_2 : PLAYER_1);
+    setGameWinner(checkForWinner());
   };
 
   return (
     <div className="App">
       <header className="App-header">
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2 id="win">Winner is... {gameWinner}</h2>
```

Comment:
```suggestion
        <h2 id="win">Winner is {gameWinner}</h2>
```
You wave 3 test failed due to the `...`. However, you got the functionality working so no worries!